---
title: "Differential transcript usage analysis of smchd1 long read dataset"
author: "Xueyi Dong"
date: "18/12/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE, message = FALSE)
```

```{r}
library(DRIMSeq)
library(limma)
```

## Loading data

Count table generated by Luyi's scFLT software

```{r loadData}
scBarcode <- read.delim("scBarcode.tsv", sep = "\t", stringsAsFactors = FALSE)
counts <- read.csv("../scFLT/run1/transcript_count.csv")
m <- match(colnames(counts)[3:ncol(counts)], scBarcode$barcode)
colnames(counts)[3:ncol(counts)] <- scBarcode$sample[m]
colnames(counts)[1] <- "feature_id"
# combine the wt sample that was sequenced twice
counts$BC07 <- counts$BC07 + counts$BC13
BC13 <- which(colnames(counts)=="BC13")
counts <- counts[,-BC13]
group <- scBarcode$group[m][-(BC13-2)]
samples <- data.frame(sample_id = colnames(counts)[3:ncol(counts)], group=group)
sel <- group != "fresia" & group != "none"
group <- group[sel]
```

```{r annotation}
library(Mus.musculus)
geneid <- as.character(counts$gene_id)
geneid <- substr(geneid, 1, 18)
genes <- select(Mus.musculus, keys=geneid, columns=c("SYMBOL", "TXCHROM","ENTREZID"), 
                keytype="ENSEMBL")
# genes <- genes[!duplicated(genes$ENSEMBL),]
m <- match(geneid, genes$ENSEMBL)
genes <- genes[m,]
head(genes)

```

generate gene level count for comparison
```{r}
countsEntrez <- counts
countsEntrez$gene_id <- genes$ENTREZID[match(geneid, genes$ENSEMBL)]
geneCount <- aggregate(countsEntrez[,c(-1,-2)], by=list(countsEntrez$gene_id), FUN=sum)
geneCount <- geneCount[,c(TRUE, sel)]
rownames(geneCount) <- geneCount[,1]
geneCount <- geneCount[,-1]
load("countsPrimary.RData")
fcCount <- fc$counts
colnames(fcCount) <- c(paste0("BC0", 7:9), paste0("BC", 10:18))
samplenames = paste("sample", c(5, 1, 6, 3, 7, 4, 2))


plotGeneCor <- function(i, g=geneCount, f=fcCount){
  m.gene = match(rownames(g), rownames(f))
  m.sample = match(colnames(g)[i], colnames(f))
  cpmf = cpm(f, log=TRUE)
  cpmg = cpm(g, log=TRUE)
  corr = cor(cpmg[,i], cpmf[m.gene, m.sample], use = "complete.obs")
  smoothScatter(cpmg[,i], cpmf[m.gene, m.sample],
                 xlab = "log CPM of gathered gene count",
                 ylab = "log CPM of direct gene count",
                 main = paste0(samplenames[i], ", cor=", round(corr, 3)))
  
}

pdf("plots/countComp.pdf", height = 4, width = 8)
par(mfrow=c(2,4))
for(i in 1:7){
  plotGeneCor(i)
  abline(coef=c(0,1), col="red")
}
dev.off()

```


```{r dmDSdata}
d <- dmDSdata(counts = cbind(counts[,c(1,2)], counts[,3:ncol(counts)][,sel]), samples = samples[sel,])
d <- dmFilter(d, min_samps_gene_expr = 7, min_samps_feature_expr = 3, min_gene_expr = 10, min_feature_expr = 10, run_gene_twice = T)
plotData(d)
```

## Precision estimation

precision: dispersion

some gene have NA precision and it will cause problem in downstream analysis

```{r}
design <- model.matrix(~group, data=DRIMSeq::samples(d))
colnames(design) <- sub("group", "", colnames(design))
set.seed(1904)
d <- suppressWarnings(dmPrecision(d, design, BPPARAM = BiocParallel::MulticoreParam(16)))

plotPrecision(d)
```

Check precision value.

Note: dispersion = 1 / (1 + precision)

```{r}
head(mean_expression(d))
common_precision(d)
head(genewise_precision(d))
```


```{r}
# look into NA dispersion genes
NA.genes <- which(is.na(genewise_precision(d)))
```


## proportion estimation

```{r}
d <- suppressWarnings(dmFit(d, design = design, verbose=1, BPPARAM = BiocParallel::MulticoreParam(16)))
head(proportions(d))
```
## testing for differential transcript usage

Test results on gene level

```{r}
d <- dmTest(d, coef = "wt")
head(results(d))
plotPValues(d)
```

Testing results on feature level (not that important for us)

```{r}
head(results(d, level = "feature"))
plotPValues(d, level = "feature")
```

explore results (gene level)

```{r}
res <- results(d)
res.txp <- results(d, level = "feature")
## turn NA p value into 1
no.na <- function(x) ifelse(is.na(x), 1, x)
res$pvalue <- no.na(res$pvalue)
res.txp$pvalue <- no.na(res.txp$pvalue)

# res <- res[order(res$pvalue, decreasing = FALSE), ]
# m <- match(substr(res$gene_id, 1, 18), genes$ENSEMBL)
# res <- cbind(res, SYMBOL=genes$SYMBOL[m])
write.table(res, "DTUres.tsv", sep = "\t", row.names = FALSE)
# top_gene_id <- res$gene_id[1]
# TO DO: change group color; change title of plot to symbol
library(RColorBrewer)
library(ggplot2)
col <- brewer.pal(3, "Set2")
# top 20 genes
pdf("plots/DTUProp.pdf", height = 6, width = 9)
for (i in 1:5){
  p <- plotProportions(d, gene_id = res$gene_id[i], group_variable = "group", 
                       plot_type = "ribbonplot", group_colors = col[1:2]) + 
    ggtitle(res$SYMBOL[i]) +
    theme(legend.text=element_text(size=5))
  print(p)
}
dev.off()

```



None of these imprinted genes are significant.

Two-stage test (code copied from DRIMSeq vignette)

```{r}
library(stageR)
## Assign gene-level pvalues to the screening stage
pScreen <- res$pvalue
# strp <- function(x) substr(x,1,18)
names(pScreen) <- results(d)$gene_id
## Assign transcript-level pvalues to the confirmation stage
pConfirmation <- matrix(res.txp$pvalue, ncol = 1)
rownames(pConfirmation) <- res.txp$feature_id
## Create the gene-transcript mapping
tx2gene <- res.txp[,c("feature_id", "gene_id")]
# for (i in 1:2) tx2gene[,i] <- strp(tx2gene[,i])
## Create the stageRTx object and perform the stage-wise analysis
stageRObj <- stageRTx(pScreen = pScreen, pConfirmation = pConfirmation,
pScreenAdjusted = FALSE, tx2gene = tx2gene)
stageRObj <- stageWiseAdjustment(object = stageRObj, method = "dtu",
alpha = 0.25)

suppressWarnings({
  drim.padj <- getAdjustedPValues(stageRObj, order=TRUE,
                                  onlySignificantGenes=TRUE)
})
head(drim.padj)
write.table(drim.padj, file = "DTUpadj.txt", sep = "\t", row.names = FALSE)
```

Only gene Pisd is significant. Let's plot proportions for Pisd.

```{r}
source("dm_plotProportions.R")

pdf("plots/DTUPropPisd.pdf", height = 6, width =9)

plotProportions(d, gene_id = "ENSMUSG00000023452.19", group_variable = "group", 
                plot_type = "ribbonplot", group_colors = col[1:2],
                order_features = FALSE) + 
  ggtitle("Pisd") +
  theme(legend.text=element_text(size=5))
  

dev.off()
```

## use diffSplice

```{r}
library(edgeR)
library(limma)
```

```{r filter}
# rownames(counts) <- counts$feature_id
# y <- DGEList(counts=counts[,c(-1,-2)])
# table(filter)
# y <- y[filter,]

# use dataset filtered by DRIMSeq
y <- DGEList(counts = as.matrix(DRIMSeq::counts(d)[,c(-1, -2)]), samples = DRIMSeq::samples(d))
rownames(y) <- counts(d)$feature_id
```

```{r MDS}
y <- calcNormFactors(y)
y$samples
cpm <- cpm(y, log=T)
plotMDS(cpm, col=as.numeric(as.factor(group)))
```

### Annotation

```{r annotation}
library(Mus.musculus)
geneid <- counts(d)$gene_id
geneid <- substr(geneid, 1, 18)
genes <- select(Mus.musculus, keys=geneid, columns=c("SYMBOL", "TXCHROM","ENTREZID"), 
                keytype="ENSEMBL")
# genes <- genes[!duplicated(genes$ENSEMBL),]
m <- match(geneid, genes$ENSEMBL)
genes <- genes[m,]
head(genes)
y$genes <- genes
```

```{r fit}
v <- voom(y, design, plot=T)
fit <- lmFit(v,design)
efit <- eBayes(fit)
dt <- decideTests(efit)
summary(dt)
```

```{r tt}
tt <- topTable(efit,  number=Inf)
```

differential exon retention

```{r}
geneid <- counts(d)$gene_id
featureid <- counts(d)$feature_id

ex <- diffSplice(efit, geneid = geneid, exonid=featureid)
ts <- topSplice(ex,  number=Inf)
```

Also only Pisd is significant.

```{r}
plotSplice(ex, geneid = "Pisd", genecolname="SYMBOL", FDR = 0.25)
```

Combine with results from Kelan's diffSplice results

```{r}
ts.short <- read.table("/wehisan/general/academic/seq_data/kelan/AGRF_CAGRF6583/diffSplice_WtvsSmchd1_allgenes_XD.txt", stringsAsFactors = FALSE, sep = "\t", header=TRUE)

calcTopN <- function(tl, ts, n){
  num <- sapply(1:n, function(x){
    length(intersect(tl[1:x, "ENTREZID"], ts[1:x, "EntrezID"]))
  }, simplify = TRUE)
  intTopN <- data.frame(
    number = 1:n,
    intersect = num
  )
  return(intTopN)
}

top200 <- calcTopN(ts, ts.short, 200)
```



# Use Gviz to plot transcripts of DTU genes

```{r}
library(Gviz)
library(GenomicFeatures)
# library(biomaRt)
# gff <- "../scFLT/run1/isoform_annotated.gff3"
gff <- "../scFLT/run1/pisd.gff3"
anno <- makeTxDbFromGFF(gff, organism = "Mus musculus")

tr_Pisd <- GeneRegionTrack(anno, chromosome = "chr5")
# ensembl <- biomaRt::useMart("ensembl")
# datasets <- biomaRt::listDatasets(ensembl)
# ensembl <- biomaRt::useDataset("mmusculus_gene_ensembl", mart=ensembl)
# biomTrack <- BiomartGeneRegionTrack(genome = "mm10", biomart = ensembl,  
#                                     name = "ENSEMBL", symbol = "Pisd",
#                                     filters=list(ensembl_transcript_id="ENSMUST00000201980")) 
gtrack <- GenomeAxisTrack()
pdf("plots/genePisd.pdf", width = 10, height = 5)
plotTracks(list(gtrack, tr_Pisd), transcriptAnnotation="transcript", lwd = 1, lwd.grid = 1)
dev.off()
``` 
