---
title: "smchd1 long read MommeD1 only analysis"
author: "Xueyi Dong"
date: "25/10/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
load("countsPrimary.RData")
# counts_run2 <- c1
# load("../Charity/counts/counts.RData")
# counts_run1 <- c1
```

```{r}
library(edgeR)
library(limma)
# source("~/analysis/voom_group/voomByGroup.R")
# x <- DGEList(counts = cbind(counts_run1$counts, counts_run2$counts))
x <- DGEList(counts = fc$counts)
colnames(x)<- c(paste0("BC0", 7:9), paste0("BC", 10:18))
x$samples$group <- factor(c("wt", "fresia", "fresia", "smchd1", "smchd1", "wt", "wt", "fresia", "wt", "wt", "smchd1", "fresia"))
x$samples$run <- factor(rep(c(1, 2), c(6,6)))
x$samples$sample <- c("Fr31", "Fr46", "Fr42", "MD143", "MD142", "MD135", "Fr31", "Fr15", "Fr17", "MD114", "MD140", "Fr46")
x$samples

# remove Fresia samples and low count samples
x <- x[, -c(2, 3, 8, 12)]

filter <- filterByExpr(x)
summary(filter)
x <- x[filter, keep.lib.sizes=FALSE]
x <- calcNormFactors(x)
lcpm <- cpm(x, log=TRUE)


x$samples
summary(x$samples$lib.size)

library(RColorBrewer)
col <- brewer.pal(3, "Set2")

# pdf("plots/mds.pdf", height = 4, width = 4)
# plotMDS(lcpm, col=col[x$samples$group], pch = c(16, 17)[x$samples$run], cex = 1.3)
# legend("topright", 
#        c("Smchd1-null run 1", "Smchd1-null run 2", "wt run 1", "wt run 2"), 
#        pch = c(16, 17, 16, 17), 
#        col = col[c(1, 1, 2, 2)], 
#        text.col = col[c(1, 1, 2, 2)],
#        pt.cex = 1.3)
# dev.off()


# plotMDS(removeBatchEffect(lcpm, batch=x$samples$run), col=col[x$samples$group], main="group")
```

Merge reads from the same sample that was sequenced twice

```{r}
# merge from run 2 to run 1
x$counts[, 1] <- x$counts[,1] + x$counts[, 5]
x$samples$lib.size[1] <- x$samples$lib.size[1] + x$samples$lib.size[5]
x$samples$run[1] <- "NA"
# delete the sample from run 2
x <- x[,-5]
x$samples
x <- calcNormFactors(x)
lcpm <- cpm(x, log=TRUE)

pdf("plots/mds_merged.pdf", height = 5, width = 7)
par(xpd=TRUE, mar =c(3.1, 4.1, 4.1, 8.1))
plotMDS(lcpm, col=col[x$samples$group], labels = 1:7, asp = 1)
# plot(1,1, type="n", yaxt="n", xaxt="n", ylab="", xlab="", frame.plot=FALSE)
legend("topright",inset=c(-0.31,0), c("Smchd1-null", "wt"), text.col = col[c(1,2)])
dev.off()
```

```{r}
pdf("plots/libsize.pdf", height = 5, width = 8)
library(ggplot2)
ggplot(x$samples, aes(x=sample, y=lib.size, fill=group)) +
  geom_bar(stat="identity") +
  scale_fill_brewer(palette = "Set2") +
  scale_x_discrete(limits = x$samples$sample, labels=1:7) +
  theme_bw()
dev.off()
```


### Annotation

```{r annotation}
library(Mus.musculus)
geneid <- rownames(x)
geneid <- substr(geneid, 1, 18)
genes <- select(Mus.musculus, keys=geneid, 
                columns=c("SYMBOL", "TXCHROM","ENSEMBL"), 
                keytype="ENTREZID")
genes <- genes[!duplicated(genes$ENTREZID),]
head(genes)
# Add gene length information, from fc
genes$length <- fc$annotation$Length[match(genes$ENTREZID, fc$annotation$GeneID)]
loglength = log(genes$length)
x$genes <- genes
```

QC: gene length VS expression (CPM)

```{r}

pdf("plots/lengthcpm.pdf", height = 4, width = 8)
par(mfrow=c(2,4))
for (i in 1:7) {
  smoothScatter(loglength, lcpm[,i], 
              xlab = "log gene length", ylab = "log CPM",
              main = paste0("Sample ", i, ", cor=", round(cor(loglength, lcpm[,i]), 3)))
}
dev.off()
```




### Design and contrast

```{r}
design <- model.matrix(~0+x$samples$group)
colnames(design) <- sub("x\\$samples\\$group", "", colnames(design))
colnames(design) <- sub("x\\$samples\\$", "", colnames(design))
cont <- makeContrasts(smchd1-wt, levels=colnames(design))
```

```{r}
# v <- voomByGroup(x, design=design, plot="all")
pdf("plots/voom.pdf", height = 4, width = 8)
v <- voomWithQualityWeights(x, design=design, plot=TRUE, col=col[x$samples$group])
dev.off()
```

### Fit

```{r}
fit <- lmFit(v, design = design)
fit <- contrasts.fit(fit, contrasts = cont)
efit <- eBayes(fit)
dt <- decideTests(efit, p.value = 0.25)
summary(dt)
```

Wow - way more DE genes when we merge the rep samples!

### Top table

```{r toptable}
# for (i in 1:3){
  # cat(colnames(efit)[i], "\n")
  print(topTable(efit, coef=1))
  # cat("\n")
# }
tt <- topTable(efit, coef=1, number=Inf)
```

length vs average expression

```{r}
library(ggplot2)
library(viridis)
pdf("plots/lengthAveExp.pdf", height = 4, width = 8)
ggplot(tt, aes(x=length, y=AveExpr)) +
  stat_binhex() +
  theme_bw() +
  labs(y = "average expression") +
  scale_x_continuous(trans = "log10") +
  scale_fill_viridis(direction = -1, option="A")
dev.off()
```


### MD plot

```{r MDplot}
pdf("plots/md.pdf")
  plotMD(efit, status=dt, column=1, values=c(1,-1), hl.col=c("red", "blue"))
dev.off()
```

### heatmap

Use top 200 genes

```{r}
library(pheatmap)
topGenes <- match(tt[1:200, 1], x$genes$ENTREZID)

annotation_col <- data.frame(group = x$samples$group)
ann_colors = list(
  group = c(smchd1 = col[1], wt = col[2])
)
rownames(annotation_col) <- colnames(x)
pdf("plots/heatmap.pdf", width = 4)
pheatmap(lcpm[topGenes,], scale = "row", 
         color = colorRampPalette(colors = c("blue","white","red"))(100),
         annotation_col = annotation_col,
         annotation_colors = ann_colors,
         show_rownames = FALSE,
         labels_col = 1:7,
         border_color = NA)
dev.off()
```

## geneset test from Chen et al. 2015
```{r}
up <- read.csv("../Charity/Chen_et_al_2015_PNAS/Upregulated_in_Smchd1_null.csv", skip=1)
down <- read.csv("../Charity/Chen_et_al_2015_PNAS/Downregulated_in_Smchd1_null.csv", skip=1)

index.up <- x$genes$ENTREZID %in% up$EntrezID
index.down <- x$genes$ENTREZID %in% down$EntrezID


plotMD(efit, status=as.numeric(index.up), column=1, main=paste0("Up genes in\n", colnames(efit)[1]))
plotMD(efit, status=as.numeric(index.down), column=1, main=paste0("Down genes in\n", colnames(efit)[1]))
```

### roast
```{r}
index.all <- index.up | index.down
gene.weights <- rep(0, nrow(x))
m <- match(up$EntrezID, x$genes$ENTREZID)
gene.weights[m[!is.na(m)]] <- up$logFC[!is.na(m)]
m <- match(down$EntrezID, x$genes$ENTREZID)
gene.weights[m[!is.na(m)]] <- down$logFC[!is.na(m)]

# for (i in 1:3) {
  cat("Test for Kelan's genes in", colnames(cont)[1], "\n")
  print(roast(v, index.all, gene.weights = gene.weights[index.all], design=design, contrast=cont[,1]))
  cat("\n")
# }

# pdf("barcode.pdf", height=12, width=10)
# png("barcode.png", height = 800, width = 640)
# par(mfrow=c(3,1))
# for(i in 1:3){
  barcodeplot(efit$t[, 1], index=index.all,
              gene.weights = gene.weights[index.all],
              main=colnames(cont)[1])
# }
# dev.off()
```

## Gene set test from Chen et al., use less genes

```{r}
up.strict <- up[up$adj.P.Val < 0.0001, ]
down.strict <- down[down$adj.P.Val < 0.0001, ]
dim(up.strict)
dim(down.strict)
index.up.strict <- x$genes$ENTREZID %in% up.strict$EntrezID
index.down.strict <- x$genes$ENTREZID %in% down.strict$EntrezID
index.strict <- index.up.strict | index.down.strict
# calculate gene weight
gene.weights <- rep(0, nrow(x))
m <- match(up.strict$EntrezID, x$genes$ENTREZID)
gene.weights[m[!is.na(m)]] <- up.strict$logFC[!is.na(m)]
m <- match(down.strict$EntrezID, x$genes$ENTREZID)
gene.weights[m[!is.na(m)]] <- down.strict$logFC[!is.na(m)]
# roast
cat("Test for Kelan's genes in", colnames(cont)[1], "\n")
print(roast(v, index.strict, gene.weights = gene.weights[index.strict], 
            design=design, contrast=cont[,1], nrot = 9999))
cat("\n")
```

Up p-value is getting higher...emmmmm

Barcode plot

```{r}
pdf("plots/barcodeStrict.pdf", height = 4)
barcodeplot(efit$t[, 1], index=index.strict,
            gene.weights = gene.weights[index.strict],
            main=colnames(cont)[1])
dev.off()
```

## Gene set test of imprinted genes

```{r}
imprinted <- read.delim("imprinted_genelist.txt")
head(imprinted)
index.imprinted <- x$genes$SYMBOL %in% imprinted$Loci
cat("Test for imprinted genes in", colnames(cont)[1], "\n")
print(roast(v, index.imprinted, design=design, contrast=cont[,1]), nrot = 9999)
cat("\n")
```

MD plot highlighting imprinted genes

```{r}
status.imprinted <- rep(0, nrow(x))
status.imprinted[x$genes$SYMBOL %in% imprinted[imprinted$Expressed.Parental.Allele == "M", 1]] <- 1
status.imprinted[x$genes$SYMBOL %in% imprinted[imprinted$Expressed.Parental.Allele == "P", 1]] <- -1
plotMD(efit, status = status.imprinted, main = "Imprinted genes")
```


barcode plot

```{r}
barcodeplot(efit$t[, 1], index=index.imprinted,
            main=colnames(cont)[1])
```


### Compare long read and old short read log FC

```{r}
tt <- topTable(efit, number=Inf)
tt.chen <- read.table("~/analysis/smchd1/old/TopTable.txt", sep="\t")
m <- match(tt$ENTREZID, tt.chen$EntrezID)
cor(tt$logFC, -tt.chen$logFC[m], use = "complete.obs")

m.up <- match(up.strict$EntrezID, tt$ENTREZID)
m.down <- match(down.strict$EntrezID, tt$ENTREZID)

smoothScatter(tt$logFC, -tt.chen$logFC[m],
              xlab = "long read log FC",
              ylab = "short read log FC",
              )
abline(coef = c(0, 1))
abline(v=0)
abline(h=0)
# highlight significant genes
points(tt$logFC[m.up], -tt.chen$logFC[m][m.up], col="red", pch = 20)
points(tt$logFC[m.down], -tt.chen$logFC[m][m.down], col="green", pch = 20)
```

The correlation is only 0.7. That's low.

Let's compare t.

```{r}
cor(tt$t, -tt.chen$t[m], use = "complete.obs")
smoothScatter(tt$t, -tt.chen$t[m],
              xlab = "long read t-statistics",
              ylab = "short read t-statistics",
              )
abline(coef = c(0, 1))
abline(v=0)
abline(h=0)
# highlight significant genes
points(tt$t[m.up], -tt.chen$t[m][m.up], col="red", pch = 20)
points(tt$t[m.down], -tt.chen$t[m][m.down], col="green", pch = 20)
```

Oh, even lower

### Compare log FC in glimma plot

```{r}
m <- match(tt$ENTREZID, tt.chen$EntrezID)
m2 <- match(tt$ENTREZID, rownames(lcpm))
rownames(x$genes) <- x$genes$ENTREZID
genes <- x$genes[m2,]
genes$long.read.adj.P.Val <- tt$adj.P.Val
genes$short.read.adj.P.Val <- tt.chen$adj.P.Val[m]
glXYPlot(tt$logFC, -tt.chen$logFC[m], counts = lcpm[m2,], launch=FALSE, 
         groups = x$samples$group,
         anno = genes,
         side.main = "SYMBOL",
         xlab = "long read logFC", ylab = "short read logFC",
         status = dt[m2,]
         )
```

See if any gene is changed to different direction in the two datasets

```{r}
up.long <- tt$ENTREZID[tt$adj.P.Val <= 0.3 & tt$logFC > 0]
down.long <- tt$ENTREZID[tt$adj.P.Val <= 0.3 & tt$logFC < 0]
up.long[up.long %in% down.strict$EntrezID]
down.long[down.long %in% up.strict$EntrezID]
```

Yes there is some.


## Save top table

```{r}
write.table(topTable(efit, coef=1, n=Inf), file = "results/TopTableMD.txt", sep="\t")
```




```{r}
sessionInfo()
```

