---
title: "Differential transcript usage analysis of smchd1 long read dataset using DRIMSeq"
author: "Xueyi Dong"
date: "19/11/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE, message = FALSE)
```

```{r}
library(DRIMSeq)
library(limma)
```

## Loading data

Count table generated by Luyi's scFLT software

```{r}
scBarcode <- read.delim("scBarcode.tsv", sep = "\t", stringsAsFactors = FALSE)
# now we only have counts from run2. Will have data from run1 soon.
counts <- read.csv("../scFLT/run2/transcript_count.csv")
m <- match(colnames(counts)[3:8], scBarcode$barcode)
colnames(counts)[3:8] <- scBarcode$sample[m]
colnames(counts)[1] <- "feature_id"
head(counts)
group <- scBarcode$group[m]
samples <- data.frame(sample_id = colnames(counts)[3:8], group=group)
sel <- group != "fresia"
group <- group[sel]
d <- dmDSdata(counts = cbind(counts[,c(1,2)], counts[,3:8][,sel]), samples = samples[sel,])
d <- dmFilter(d, min_samps_gene_expr = 2, min_samps_feature_expr = 2, min_gene_expr = 10, min_feature_expr = 5, run_gene_twice = T)
plotData(d)
```



## Precision estimation

precision: dispersion

some gene have NA precision and it will cause problem in downstream analysis

```{r}
design <- model.matrix(~group, data=samples(d))
colnames(design) <- sub("group", "", colnames(design))
set.seed(1904)
d <- suppressWarnings(dmPrecision(d, design, BPPARAM = BiocParallel::MulticoreParam(16)))

plotPrecision(d)
```

Check precision value.

Note: dispersion = 1 / (1 + precision)

```{r}
head(mean_expression(d))
common_precision(d)
head(genewise_precision(d))
```


```{r}
# look into NA dispersion genes
NA.genes <- which(is.na(genewise_precision(d)))
```


## proportion estimation

```{r}
d <- suppressWarnings(dmFit(d, design = design, verbose=1, BPPARAM = BiocParallel::MulticoreParam(16)))
head(proportions(d))
```
## testing for differential transcript usage

Test results on gene level

```{r}
d <- dmTest(d, coef = "groupwt")
head(results(d))
plotPValues(d)
```

Testing results on feature level

```{r}
head(results(d, level = "feature"))
plotPValues(d, level = "feature")
```

## Two-stage test

Error exists for adjustment. 
features of NA p-values manually removed. set allowNA=TRUE doesn't work and I don't know why!

```{r, eval=FALSE}
library(stageR)
## Assign gene-level pvalues to the screening stage
pScreen <- results(d2)$pvalue
names(pScreen) <- results(d2)$gene_id
naFeatures <- which(is.na(pScreen))
pScreen <- pScreen[-naFeatures]
## Assign transcript-level pvalues to the confirmation stage
pConfirmation <- results(d2, level = "feature")
naFeatures <- which(pConfirmation$gene_id %in% results(d2)[naFeatures,"gene_id"])
pConfirmation <- matrix(pConfirmation[-naFeatures, "pvalue"], ncol = 1)
rownames(pConfirmation) <- results(d2, level = "feature")$feature_id[-naFeatures]
## Create the gene-transcript mapping
tx2gene <- results(d2, level = "feature")[-naFeatures, c("feature_id", "gene_id")]
## Create the stageRTx object and perform the stage-wise analysis
stageRObj <- stageRTx(pScreen = pScreen, pConfirmation = pConfirmation, pScreenAdjusted = FALSE, tx2gene = tx2gene)
stageRObj <- stageWiseAdjustment(object = stageRObj, method = "dtu", alpha = 0.05)
getSignificantGenes(stageRObj)
getSignificantTx(stageRObj)
padj <- getAdjustedPValues(stageRObj, order = T, onlySignificantGenes = FALSE)
head(padj)
```

```{r, eval=FALSE}
library(stageR)
pScreen <- results(d1)$pvalue
names(pScreen) <- results(d1)$gene_id
## Assign transcript-level pvalues to the confirmation stage
pConfirmation <- matrix(results(d1, level = "feature")$pvalue, ncol = 1)
rownames(pConfirmation) <- results(d1, level = "feature")$feature_id
## Create the gene-transcript mapping
tx2gene <- results(d1, level = "feature")[, c("feature_id", "gene_id")]
## Create the stageRTx object and perform the stage-wise analysis
stageRObj <- stageRTx(pScreen = pScreen, pConfirmation = pConfirmation, pScreenAdjusted = FALSE, tx2gene = tx2gene)
stageRObj <- stageWiseAdjustment(object = stageRObj, method = "dtu", alpha = 0.05, allowNA=TRUE)
getSignificantGenes(stageRObj)
getSignificantTx(stageRObj)
padj <- getAdjustedPValues(stageRObj, order = TRUE, onlySignificantGenes = FALSE)
head(padj)
```

## use diffSplice

```{r}
library(edgeR)
library(limma)
```

```{r filter}
rownames(counts) <- counts$feature_id
y <- DGEList(counts=counts[,c(-1,-2)])
filter <- filterByExpr(y, design = design)
table(filter)
y <- y[filter,]
```

```{r MDS}
y <- calcNormFactors(y)
y$samples
cpm <- cpm(y, log=T)
plotMDS(cpm, col=as.numeric(group))
```

### Annotation

```{r annotation}
library(Mus.musculus)
geneid <- counts$gene_id[filter]
geneid <- substr(geneid, 1, 18)
genes <- select(Mus.musculus, keys=geneid, columns=c("SYMBOL", "TXCHROM","ENTREZID"), 
                keytype="ENSEMBL")
# genes <- genes[!duplicated(genes$ENSEMBL),]
m <- match(geneid, genes$ENSEMBL)
genes <- genes[m,]
head(genes)
y$genes <- genes
```

```{r fit}
contr <- makeContrasts(fresia-wt, smchd1-wt, fresia-smchd1, levels=colnames(design))
v <- voom(y, design, plot=T)
fit <- lmFit(v,design)
fit <- contrasts.fit(fit, contr)
efit <- eBayes(fit)
dt <- decideTests(efit)
summary(dt)
```

```{r tt}
tt_f_wt <- topTable(efit, coef=1, number=Inf)
tt_s_wt <- topTable(efit, coef=2, number=Inf)
tt_f_s <- topTable(efit, coef=3, number=Inf)
```

differential exon retention

```{r}
geneid <- counts[filter,1]
featureid <- counts[filter,2]

ex <- diffSplice(fit, geneid = geneid, exonid=featureid)
ts_f_wt <- topSplice(ex, coef=1, number=Inf)
ts_s_wt <- topSplice(ex, coef=2, number=Inf)
ts_f_s <- topSplice(ex, coef=3, number=Inf)
write.table(ts_f_wt, file="dtu/topSplice_fresia-wt.txt", sep="\t")
write.table(ts_s_wt, file="dtu/topSplice_smchd1-wt.txt", sep="\t")
write.table(ts_f_s, file="dtu/topSplice_fresia-smchd1.txt", sep="\t")
```

gene level test
```{r}
tsg_f_wt <- topSplice(ex, coef=1, number=Inf, test="F")
tsg_s_wt <- topSplice(ex, coef=2, number=Inf, test="F")
tsg_f_s <- topSplice(ex, coef=3, number=Inf, test="F")
write.table(tsg_f_wt, file="dtu/topSplice_G_fresia-wt.txt", sep="\t")
write.table(tsg_s_wt, file="dtu/topSplice_G_smchd1-wt.txt", sep="\t")
write.table(tsg_f_s, file="dtu/topSplice_G_fresia-smchd1.txt", sep="\t")
```

### number of significant genes

```{r}
nSigGene <- rbind(nSigGene, c(
  "diffSplice", "fresiaVSwt",
  nrow(tsg_f_wt[tsg_f_wt$FDR<0.05,])
), c(
  "diffSplice", "smchd1VSwt",
  nrow(tsg_s_wt[tsg_s_wt$FDR<0.05,])
), c(
  "diffSplice", "fresiaVSsmchd1",
  nrow(tsg_f_s[tsg_f_s$FDR<0.05,])
))

nSigGene$n <- as.numeric(nSigGene$n)
```


```{r}
res_merge_1 <- merge(sres1, tsg_f_wt, by = "GeneID", suffixes = c(".short", ".long"))
# head(res_merge_1)
res_merge_2 <- merge(sres2, tsg_s_wt, by = "GeneID", suffixes = c(".short", ".long"))
res_merge_3 <- merge(sres3, tsg_f_s, by = "GeneID", suffixes = c(".short", ".long"))
```

```{r}
library(tidyverse)

res_sig1 <- res_merge_1 %>% filter(P.Value.short <= 0.05) %>% filter(P.Value.long <= 0.05)
res_sig1

res_sig2 <- res_merge_2 %>% filter(P.Value.short <= 0.05) %>% filter(P.Value.long <= 0.05)
res_sig2

res_sig3 <- res_merge_3 %>% filter(P.Value.short <= 0.05) %>% filter(P.Value.long <= 0.05)
res_sig3
```

## Bar plot of differencial genes found by each method

```{r}
library(ggplot2)
ggplot(nSigGene, aes(x=contrast, y=n, fill=method)) + 
  geom_bar(stat="identity", position=position_dodge())
```

