---
title: "DTU analysis of sequins data"
author: "Xueyi Dong"
date: "20/03/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(limma)
library(edgeR)
library(data.table)
library(ggplot2)
```

First, read in the data

```{r}
data <- read.csv("/stornext/General/data/user_managed/grpu_mritchie_1/SCmixology/Mike_seqin/20200228_YPRDP_2xsequin_mixAB/flames/transcript_count.csv", stringsAsFactors = FALSE)
anno <- read.delim("/stornext/General/data/user_managed/grpu_mritchie_1/SCmixology/Mike_seqin/annotations/rnasequin_isoforms_2.4.tsv", sep = "\t", stringsAsFactors = FALSE)
# tracking <- read.table("comp.tracking", stringsAsFactors = FALSE)
# info <- strsplit2(tracking$V5, "|", fixed=TRUE)
# tracking$transcript <- sub("transcript:", "", info[,2])
# tracking$transcript_anno <- strsplit2(tracking$V3, "|", fixed=TRUE)[,2]
# tracking.eq <- tracking[tracking$V4=="=",]
# m <- match(data$transcript_id, tracking.eq$transcript)
# data$transcript_anno <- tracking.eq$transcript_anno[m]
# data$transcript_anno[is.na(data$transcript_anno)] <- data$transcript_id[is.na(data$transcript_anno)]
```



```{r}
# m <- match(anno$NAME, data$transcript_id)
# cor(data[,3][m], anno[,3], us="complete.obs")
# cor(data[,3][m], anno[,4], us="complete.obs")
# cor(data[,4][m], anno[,3], us="complete.obs")
# cor(data[,4][m], anno[,4], us="complete.obs")
# cor(data[,5][m], anno[,3], us="complete.obs")
# cor(data[,5][m], anno[,4], us="complete.obs")
# cor(data[,6][m], anno[,3], us="complete.obs")
# cor(data[,6][m], anno[,4], us="complete.obs")
```


```{r}
# par(mfrow=c(2,2))
# plot(data[,3][m], anno[,4], log="xy", 
#      xlab = "count", ylab="theoretiacl abundance", main="sample 1, mix B")
# plot(data[,4][m], anno[,4], log="xy", 
#      xlab = "count", ylab="theoretiacl abundance", main="sample 2, mix B")
# plot(data[,5][m], anno[,3], log="xy", 
#      xlab = "count", ylab="theoretiacl abundance", main="sample 3, mix A")
# plot(data[,6][m], anno[,3], log="xy", 
#      xlab = "count", ylab="theoretiacl abundance", main="sample 4, mix A")
```

```{r}
cpm <- cpm(data[3:6])
m <- match(anno$NAME, data$transcript_id)
annoData <- cbind(anno, cpm[m,])
colnames(annoData)[5:8] <- paste0("mix", c("B1", "A2", "B2", "A1"))
annoData <- melt(annoData, id.vars = 1:4)
annoData$exp <-annoData$MIX_A
annoData$exp[annoData$variable %in% c("mixB1", "mixB2")] <- annoData$MIX_B[annoData$variable %in% c("mixB1", "mixB2")]
colnames(annoData)[5] <- "sample"
annoData$sample <- factor(annoData$sample, levels = c("mixA1", "mixA2", "mixB1", "mixB2"))
cor(annoData$exp, annoData$value, use = "complete.obs")

library(ggplot2)
library(scales)
pdf("plots/countsVsAnno.pdf", height = 5, width = 8)
ggplot(annoData, aes(x=exp, y=value, colour=sample)) +
  geom_point() +
  theme_bw() +
  scale_x_continuous(trans = "log10", labels = comma) +
  scale_y_continuous(trans = "log10", labels=comma, limits=c(0.5, 180000)) +
  labs(x = "Expected abundance", y = "CPM") +
  scale_colour_manual(values = c("#869700", "#D12300", "#071535", "#54B5E1"))+
  theme(text = element_text(size = 20)) +
  annotate(geom="text", x=300, y=160000, label=expression(paste(rho, "=0.81")), size=8) 
dev.off()

# lm.count <- lm(annoData$value~annoData$exp)
# summary(lm.count)
# cor(annoData$value, annoData$exp, use = "complete.obs")
```


OK. Now prepare for DRIMSeq.

```{r}
library(DRIMSeq)
counts <- data[,c(2, 1, 3, 4, 5, 6)]
colnames(counts)[2] <- "feature_id"
samples <- data.frame(
  sample_id = colnames(counts)[3:6],
  group = c("B", "A", "B", "A")
)
# # resolve duplicate
# for(i in which(duplicated(counts$feature_id))){
#   j = which(counts$feature_id == counts$feature_id[i])[1]
#   counts[j,3:6] <- counts[j, 3:6] + counts[i, 3:6]
# }
# counts <- counts[!duplicated(counts$feature_id),]

d <- dmDSdata(counts=counts, samples = samples)
plotData(d)
d <- dmFilter(d, min_samps_gene_expr = 4, min_samps_feature_expr = 2, min_gene_expr = 10, min_feature_expr = 10, run_gene_twice = T)
plotData(d)
```

```{r}
design <- model.matrix(~group, data=DRIMSeq::samples(d))
colnames(design) <- sub("group", "", colnames(design))
set.seed(1904)
d <- suppressWarnings(dmPrecision(d, design, BPPARAM = BiocParallel::MulticoreParam(16)))

plotPrecision(d)
```

Check precision value.

Note: dispersion = 1 / (1 + precision)

```{r}
head(mean_expression(d))
common_precision(d)
head(genewise_precision(d))
```


## proportion estimation

```{r}
d <- suppressWarnings(dmFit(d, design = design, verbose=1, BPPARAM = BiocParallel::MulticoreParam(16)))
head(proportions(d))
```
## testing for differential transcript usage

Test results on gene level

```{r}
d <- dmTest(d, coef = "B")
head(results(d))
plotPValues(d)
```

Testing results on feature level (not that important for us)

```{r}
head(results(d, level = "feature"))
plotPValues(d, level = "feature")
```

Two-stage test (code copied from DRIMSeq vignette)

```{r}
res <- results(d)
res.txp <- results(d, level = "feature")
no.na <- function(x) ifelse(is.na(x), 1, x) 
res$pvalue <- no.na(res$pvalue) 
res.txp$pvalue <- no.na(res.txp$pvalue)

library(stageR)
## Assign gene-level pvalues to the screening stage
pScreen <- res$pvalue
# strp <- function(x) substr(x,1,18)
names(pScreen) <- results(d)$gene_id
## Assign transcript-level pvalues to the confirmation stage
pConfirmation <- matrix(res.txp$pvalue, ncol = 1)
rownames(pConfirmation) <- res.txp$feature_id
## Create the gene-transcript mapping
tx2gene <- res.txp[,c("feature_id", "gene_id")]
# for (i in 1:2) tx2gene[,i] <- strp(tx2gene[,i])
## Create the stageRTx object and perform the stage-wise analysis
stageRObj <- stageRTx(pScreen = pScreen, pConfirmation = pConfirmation,
pScreenAdjusted = FALSE, tx2gene = tx2gene)
stageRObj <- stageWiseAdjustment(object = stageRObj, method = "dtu",
alpha = 0.05)

suppressWarnings({
  drim.padj <- getAdjustedPValues(stageRObj, order=TRUE,
                                  onlySignificantGenes=FALSE)
})
head(drim.padj)
# write.table(drim.padj, file = "DTUpadj.txt", sep = "\t", row.names = FALSE)
```

drimseq proportion vs real proportion

calculate annotation proportion.
```{r}
geneanno <- read.delim("/stornext/General/data/user_managed/grpu_mritchie_1/SCmixology/Mike_seqin/annotations/rnasequin_genes_2.4.tsv", sep = "\t", stringsAsFactors = FALSE)
proportion <- data.frame(txID = anno$NAME)
tx2gene <- limma::strsplit2(proportion$txID, "_")
tx2gene <-  paste(tx2gene[,1], tx2gene[,2], sep="_")
proportion$geneID <- tx2gene
m <- match(proportion$geneID, geneanno$NAME)
proportion$MIX_A <- anno$MIX_A / geneanno$MIX_A[m]
proportion$MIX_B <- anno$MIX_B / geneanno$MIX_B[m]

m <- match(proportion$txID, proportions(d)[,2])
cor(proportion$MIX_B, proportions(d)[,3][m], use="complete.obs")
cor(proportion$MIX_A, proportions(d)[,4][m], use="complete.obs")
```

```{r}
par(mfrow=c(1,2))
plot(proportion$MIX_B, proportions(d)[,3][m], main="mix B",
     xlab = "expected proportion", ylab = "estimated proportion")
plot(proportion$MIX_A, proportions(d)[,4][m], main="mix A",
     xlab = "expected proportion", ylab = "estimated proportion")
```

run diffSplice

```{r}
library(edgeR)
library(limma)
y <- DGEList(counts = as.matrix(DRIMSeq::counts(d)[,c(-1, -2)]), samples = DRIMSeq::samples(d))
rownames(y) <- counts(d)$feature_id
y <- calcNormFactors(y)
y$samples
cpm <- cpm(y, log=T)
plotMDS(cpm, col=as.numeric(as.factor(y$samples$group)))
```

```{r fit}
v <- voom(y, design, plot=T)
fit <- lmFit(v,design)
efit <- eBayes(fit)
dt <- decideTests(efit, p.value = 0.25)
summary(dt)
```

```{r}
geneid <- counts(d)$gene_id
featureid <- counts(d)$feature_id

ex <- diffSplice(efit, geneid = geneid, exonid=featureid)
ts <- topSplice(ex,  number=Inf)
ts.tx <- topSplice(ex, test = "t", number = Inf)
ts.f <- topSplice(ex, test = "F", number = Inf)

head(ts)
table(ts$FDR < 0.05)
```


Two-stage test for diffSplice

```{r}

library(stageR)

pScreen <- ts$P.Value
pScreen2 <- ts.f$P.Value
names(pScreen) <- ts$GeneID
names(pScreen2) <- ts.f$GeneID
pConfirmation <- matrix(ts.tx$P.Value, ncol = 1)
rownames(pConfirmation) <- ts.tx$ExonID

tx2gene <- ts.tx[,1:2]

stageRObj2 <- stageRTx(pScreen = pScreen, pConfirmation = pConfirmation,
pScreenAdjusted = FALSE, tx2gene = tx2gene)
stageRObj2 <- stageWiseAdjustment(object = stageRObj2, method = "dtu",
alpha = 0.05)
stageRObj3 <- stageRTx(pScreen = pScreen2, pConfirmation = pConfirmation,
pScreenAdjusted = FALSE, tx2gene = tx2gene)
stageRObj3 <- stageWiseAdjustment(object = stageRObj3, method = "dtu",
alpha = 0.05)
suppressWarnings({
  ds.padj <- getAdjustedPValues(stageRObj2, order=TRUE,
                                  onlySignificantGenes=FALSE)
  ds.padj.f <- getAdjustedPValues(stageRObj3, order=TRUE,
                                  onlySignificantGenes=FALSE)
})
head(ds.padj)
head(ds.padj.f)

```

Now deal with anno, find out control genes

```{r}
table(anno$MIX_A==anno$MIX_B)
proportion$FC <- proportion$MIX_A / proportion$MIX_B
proportion$isChanges <- FALSE
proportion$isChanges[abs(proportion$FC-1) > 0.001] <- TRUE
```

check DRIMSeq results

```{r}
# drimseq feature level
res.tx <- results(d, level = "feature")
res.tx$isChanges <- proportion$isChanges[match(res.tx$feature_id, proportion$txID)]
# drimseq gene level
res.gene <- results(d, level = "gene")
res.gene$isChanges <- proportion$isChanges[match(res.gene$gene_id, proportion$geneID)]
# set stager transcript level NA to 1ï¼Ÿ no
suppressWarnings({
  drim.padj <- getAdjustedPValues(stageRObj, order=TRUE,
                                  onlySignificantGenes=FALSE)
})
# drim.padj$transcript[is.na(drim.padj$transcript)] <- 1
drim.padj$isChanges <- proportion$isChanges[match(drim.padj$txID, proportion$txID)]
a <- match(unique(drim.padj$geneID), drim.padj$geneID)
res.all <- data.frame(
  id = c(res.gene$gene_id, res.tx$feature_id, unique(drim.padj$geneID), drim.padj$txID),
  FDR = c(res.gene$adj_pvalue, res.tx$adj_pvalue, drim.padj$gene[a], drim.padj$transcript),
  DTU = c(res.gene$isChanges, res.tx$isChanges, drim.padj$isChanges[a], drim.padj$isChanges),
  test = c(rep("DRIMSeq_gene", nrow(res.gene)), rep("DRIMSeq_feature", nrow(res.tx)), rep("DRIMSeq_stageR_gene", length(unique(drim.padj$geneID))), rep("DRIMSeq_stageR_transcript", nrow(drim.padj))),
  level = c(rep("Gene level", nrow(res.gene)), rep("Transcript level", nrow(res.tx)), rep("Gene level", length(unique(drim.padj$geneID))), rep("Transcript level", nrow(drim.padj)))
)

ggplot(na.omit(res.all), aes(x=DTU, y=FDR, fill=test)) +
  geom_boxplot() +
  theme_bw()
```

```{r}
res.tx$logFC <-log( proportion$FC[match(res.tx$feature_id, proportion$txID)])
ggplot(res.tx, aes(x=logFC, y=adj_pvalue)) +
  geom_point() +
  theme_bw() +
  ggtitle("DRIMSeq")
```


And check diffSplice results

```{r}
ts.tx$isChanges <- proportion$isChanges[match(ts.tx$ExonID, proportion$txID)]
ggplot(na.omit(ts.tx), aes(x=isChanges, y=FDR)) +
  geom_boxplot() +
  theme_bw() +
  ggtitle("diffSplice transcript-level FDR") +
  labs(x = "isProportionChange")
  
b <- match(unique(ds.padj$geneID), ds.padj$geneID)
c <- match(unique(ds.padj.f$geneID), ds.padj.f$geneID)
ds.all <- data.frame(
  id = c(ts$GeneID, ts.f$GeneID, ts.tx$ExonID, 
         unique(ds.padj$geneID), ds.padj$txID
         # , 
         # unique(ds.padj.f$geneID), ds.padj.f$txID
         ),
  FDR = c(ts$FDR, ts.f$FDR, ts.tx$FDR, 
          ds.padj$gene[b], ds.padj$transcript
          # , 
          # ds.padj.f$gene[c], ds.padj.f$transcript
          ),
  DTU = c(proportion$isChanges[match(ts$GeneID, proportion$geneID)], 
          proportion$isChanges[match(ts.f$GeneID, proportion$geneID)],
          proportion$isChanges[match(ts.tx$ExonID, proportion$txID)], 
          proportion$isChanges[match(ds.padj$geneID, proportion$geneID)][b], 
          proportion$isChanges[match(ds.padj$txID, proportion$txID)]
          # , proportion$isChanges[match(ds.padj.f$geneID, proportion$geneID)][c], 
          # proportion$isChanges[match(ds.padj.f$txID, proportion$txID)]
          ),
  test = c(rep("diffSplice_simes", nrow(ts)), rep("diffSplice_F", nrow(ts.f)), rep("diffSplice_t", nrow(ts.tx)),
           rep("diffSplice_stageR_gene", length(unique(ds.padj$geneID))),
           rep("diffSplice_stageR_transcript", nrow(ds.padj))
           # ,
           # rep("diffSplice_F_stageR_gene", length(unique(ds.padj.f$geneID))),
           # rep("diffSplice_F_stageR_transcript", nrow(ds.padj.f))
           ),
  level = c(rep("Gene level", nrow(ts)), rep("Gene level", nrow(ts.f)), rep("Transcript level", nrow(ts.tx)),
            rep("Gene level", length(unique(ds.padj$geneID))),
            rep("Transcript level", nrow(ds.padj))
            # ,
            # rep("Gene level", length(unique(ds.padj.f$geneID))),
            # rep("Feature level", nrow(ds.padj.f))
            )
)

ggplot(na.omit(ds.all), aes(x=DTU, y=FDR, fill=test)) +
  geom_boxplot() +
  theme_bw()
```

```{r}
results.all <- rbind(res.all, ds.all)
results.all$test <- factor(results.all$test, levels = c("DRIMSeq_gene", "DRIMSeq_feature", "DRIMSeq_stageR_gene", "DRIMSeq_stageR_transcript",
                              "diffSplice_simes", "diffSplice_F", "diffSplice_t", "diffSplice_stageR_gene", "diffSplice_stageR_transcript"))

pdf("plots/DTU_FDR.pdf", width = 8, height = 5)
ggplot(na.omit(results.all), aes(x = DTU, y = FDR, fill=test, colour=test)) +
  geom_boxplot(alpha=.4, outlier.shape = NA)+
  geom_point(position = position_jitterdodge(jitter.height = .02), cex=.5, alpha = .5)+
  facet_grid(.~level) +
  theme_bw() +
  # theme(axis.text.x = element_text(angle = 90)) +
  theme(text = element_text(size=16), legend.text = element_text(size=10))+
  geom_hline(yintercept = 0.05, linetype="dashed") +
  labs(y = "Adjusted p-value", x="is DTU") +
  scale_fill_brewer(palette = "Set1")+
  scale_colour_brewer(palette = "Set1")
dev.off()
```

 


New plot for the paper

```{r}
results.plot <- results.all[results.all$test %in% c("DRIMSeq_feature", "DRIMSeq_gene", "DRIMSeq_stageR_gene", "DRIMSeq_stageR_transcript", "diffSplice_simes", "diffSplice_stageR_gene", "diffSplice_stageR_transcript", "diffSplice_t"), ]
results.plot$test <- factor(results.plot$test, levels = c("DRIMSeq_gene", "DRIMSeq_feature", "DRIMSeq_stageR_gene", "DRIMSeq_stageR_transcript", "diffSplice_simes", "diffSplice_t", "diffSplice_stageR_gene", "diffSplice_stageR_transcript" ))



pdf("plots/DTU_FDR_TRUE.pdf", height = 5, width = 8)
ggplot(na.omit(results.plot[results.plot$DTU == TRUE,]), aes(x = test, y = FDR, fill=level, colour = level))+
  geom_boxplot(alpha = .6) +
  geom_hline(yintercept = 0.05, linetype="dashed") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
        text = element_text(size = 20),
        plot.margin = unit(c(5.5, 5.5, 5.5, 25), "points"))  +
  labs(y = "Adjusted p-value", x="Method", fill = "Test level", colour = "Test level") +
  scale_fill_manual(values = c("#F09506", "#2A557B")) +
  scale_colour_manual(values = c("#F09506", "#2A557B"))
dev.off()
```


Calculate FDR and TPR.

FDR = number of falsely detected genes / number of detected genes
TPR = number of correctly detected genes / number of positive control genes

```{r}
results.all$sig <- results.all$FDR < 0.05
FDR <- sapply(unique(results.all$test), function(x){
  results = na.omit(results.all[results.all$test == x, ])
  fd = sum(results$sig & !(results$DTU))
  dt = sum(results$sig)
  return(c(as.character(x), fd/dt))
}, simplify = TRUE)
FDR

TPR <- sapply(unique(results.all$test), function(x){
  results = na.omit(results.all[results.all$test == x, ])
  td = sum(results$sig & results$DTU)
  pc = sum(results$DTU)
  return(c(as.character(x), td/pc))
}, simplify = TRUE)
TPR <- as.data.frame(t(TPR))
TPR$V2 <- as.numeric(as.character(TPR$V2))
colnames(TPR) <- c("test", "TPR")
TPR$test <- factor(TPR$test, levels = c("DRIMSeq_gene", "DRIMSeq_feature", "DRIMSeq_stageR_gene", "DRIMSeq_stageR_transcript", "diffSplice_simes", "diffSplice_F", "diffSplice_t", "diffSplice_stageR_gene", "diffSplice_stageR_transcript" ))
```

FDR all 0. now plot TPR

```{r}
pdf("plots/DTU_TPR.pdf")
ggplot(TPR, aes(x=test, y=TPR, fill=test)) +
  geom_bar(stat = "identity") +
  theme_bw() +
  labs (x = "Test", y = "True positive rate") +
  scale_fill_brewer(palette = "Set1")+
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1), text = element_text(size=16),
        legend.position = "none")
dev.off()
```


```{r}
ts.tx$logFC <- log(proportion$FC[match(ts.tx$ExonID, proportion$txID)])

ggplot(ts.tx, aes(x=logFC, y=FDR)) +
  geom_point() +
  theme_bw() +
  ggtitle("diffSplice")
```


```{r}
save.image("DTUsequins.RData")
```

## Compare results with Illumina

Combine with results from short read results.

First, calculate the overlap of top N genes of diffSplice simes results

```{r}
ts.short <- read.table("illumina/topSplice_simes.tsv", stringsAsFactors = FALSE, sep = "\t", header=TRUE)

calcTopN <- function(tl, ts, n){
  num <- sapply(1:n, function(x){
    length(intersect(tl[1:x, "GeneID"], ts[1:x, "GeneID"]))
  }, simplify = TRUE)
  intTopN <- data.frame(
    number = 1:n,
    intersect = num
  )
  return(intTopN)
}

top47 <- calcTopN(ts, ts.short, 47)
ggplot(top47, aes(x=number, y=intersect)) + geom_line() +
  geom_abline(colour="grey")
```

wow, nice!


Then, compare this of DRIMSeq gene-level results
```{r}
res.short <- read.table("illumina/drimseq_gene.tsv", stringsAsFactors = FALSE, sep = "\t", header = TRUE)
m <- match(res.short$gene_id, res$gene_id)
cor(res[m, "lr"], res.short$lr, use = "complete.obs")
cor(res[m, "pvalue"], res.short$pvalue, use = "complete.obs")
```

not comparable

and compare DRIMSeq estimated proportion
```{r}
proportions.short <- read.table("illumina/drimseq_proportion.tsv", stringsAsFactors = FALSE, sep = "\t", header = TRUE)
# only compare wt proportion
m <- match(proportions.short$feature_id, proportions(d)$feature_id)
cor(proportions(d)[m, 3], proportions.short$X6.HCC.1_S6, use = "complete.obs")
plot(proportions(d)[m, 3], proportions.short$X6.HCC.1_S6)
```

Compare long and short read DRIMSeq proportions to annotated proportions
```{r}

```



and also need to do a count vs anno plot for short read.
DTU genes venn diagram for long vs short vs downsample?