---
title: "sequins DE"
author: "Xueyi Dong"
date: "27/04/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(rsubread)
library(limma)
library(edgeR)
library(ggplot2)
```

## Counting

```{r}
bams <- paste0("./talon/barcode0", 1:4, ".sam")
gff <- "/wehisan/home/allstaff/d/dong.x/annotation/sequins/rnasequin_annotation_2.4.gtf"
fc <- featureCounts(bams, annot.ext=gff, isGTFAnnotationFile = TRUE, isLongRead = TRUE, primaryOnly=TRUE)
```

```{r}
fc$stat
```

## DE analysis

```{r}
x <- DGEList(counts = fc$counts)
x$samples$group <- c("A", "A", "B", "B")

filter <- filterByExpr(x)
summary(filter)
x <- x[filter, keep.lib.sizes=FALSE]
x <- calcNormFactors(x)
lcpm <- cpm(x, log=TRUE)
```

7 genes were filtered out. 69 remained.

```{r}
plotMDS(lcpm)
```

```{r}
design <- model.matrix(~x$samples$group)
colnames(design) <- sub("x\\$samples\\$group", "", colnames(design))
colnames(design) <- sub("x\\$samples\\$", "", colnames(design))
pdf("plots/voom_sequins.pdf", height = 4, width = 8)
v <- voomWithQualityWeights(x, design=design, plot=TRUE)
dev.off()
```

The voom trend looks OK.

```{r}
fit <- lmFit(v, design = design)
efit <- eBayes(fit)
dt <- decideTests(efit)
summary(dt)
```

21 down and 18 up in BvsA.

```{r}
plotMD(efit, status=dt[,2], column = 2, values=c(1,-1), hl.col=c("red", "blue"),
       main = "mix B vs A")
```


```{r}
tt <- topTable(efit, coef=2, number=Inf)
anno <- read.table("/wehisan/home/allstaff/d/dong.x/annotation/sequins/rnasequin_genes_2.4.tsv", header = TRUE, stringsAsFactors = FALSE)
anno$logFC <- log(anno$MIX_B / anno$MIX_A)
tt$logFC_expected <- anno$logFC[match(rownames(tt), anno$NAME)]
tt$LENGTH <- anno$LENGTH[match(rownames(tt), anno$NAME)]
cor(tt$logFC, tt$logFC_expected)
pdf("logfc.pdf", height = 5, width = 8)
ggplot(tt, aes(x = logFC_expected, y = logFC)) +
  geom_point(aes(colour=LENGTH)) +
  theme_bw()
dev.off()

```

logFC showed good correlation. GOOOOOD!

```{r}
pdf("lengthAveExpSequins.pdf", height = 4, width = 8)
ggplot(tt, aes(x = LENGTH, y = AveExpr)) +
  geom_point() +
  theme_bw() +
  labs(x = "length", y = "average expression") +
  scale_x_continuous(trans = "log10")
dev.off()
```

```{r}
table(prediction=(tt$adj.P.Val<0.05), trueLabel=tt$logFC_expected!=0)
```

Calculate FDR and TPR

```{r}
tt$sig <- tt$adj.P.Val<0.05
tt$DE <- abs(tt$logFC_expected) > 0.0001

FDR = sum(tt$sig &!(tt$DE)) / sum(tt$sig)

TPR = sum(tt$DE & tt$sig) / sum(tt$DE)
```


```{r}
#         ==========     _____ _    _ ____  _____  ______          _____  
#         =====         / ____| |  | |  _ \|  __ \|  ____|   /\   |  __ \ 
#           =====      | (___ | |  | | |_) | |__) | |__     /  \  | |  | |
#             ====      \___ \| |  | |  _ <|  _  /|  __|   / /\ \ | |  | |
#               ====    ____) | |__| | |_) | | \ \| |____ / ____ \| |__| |
#         ==========   |_____/ \____/|____/|_|  \_\______/_/    \_\_____/
#        Rsubread 1.34.7
# 
# //========================== featureCounts setting ===========================\\
# ||                                                                            ||
# ||             Input files : 4 SAM files                                      ||
# ||                           o barcode01.sam                                  ||
# ||                           o barcode02.sam                                  ||
# ||                           o barcode03.sam                                  ||
# ||                           o barcode04.sam                                  ||
# ||                                                                            ||
# ||              Annotation : rnasequin_annotation_2.4.gtf (GTF)               ||
# ||      Dir for temp files : .                                                ||
# ||                 Threads : 1                                                ||
# ||                   Level : meta-feature level                               ||
# ||              Paired-end : no                                               ||
# ||      Multimapping reads : counted                                          ||
# ||     Multiple alignments : primary alignment only                           ||
# || Multi-overlapping reads : not counted                                      ||
# ||   Min overlapping bases : 1                                                ||
# ||          Long read mode : yes                                              ||
# ||                                                                            ||
# \\============================================================================//
# 
# //================================= Running ==================================\\
# ||                                                                            ||
# || Load annotation file rnasequin_annotation_2.4.gtf ...                      ||
# ||    Features : 1175                                                         ||
# ||    Meta-features : 76                                                      ||
# ||    Chromosomes/contigs : 1                                                 ||
# ||                                                                            ||
# || Process SAM file barcode01.sam...                                          ||
# ||    Single-end reads are included.                                          ||
# ||    Total alignments : 2164910                                              ||
# ||    Successfully assigned alignments : 1812361 (83.7%)                      ||
# ||    Running time : 0.59 minutes                                             ||
# ||                                                                            ||
# || Process SAM file barcode02.sam...                                          ||
# ||    Single-end reads are included.                                          ||
# ||    Total alignments : 856806                                               ||
# ||    Successfully assigned alignments : 729141 (85.1%)                       ||
# ||    Running time : 0.27 minutes                                             ||
# ||                                                                            ||
# || Process SAM file barcode03.sam...                                          ||
# ||    Single-end reads are included.                                          ||
# ||    Total alignments : 1738360                                              ||
# ||    Successfully assigned alignments : 1266155 (72.8%)                      ||
# ||    Running time : 0.49 minutes                                             ||
# ||                                                                            ||
# || Process SAM file barcode04.sam...                                          ||
# ||    Single-end reads are included.                                          ||
# ||    Total alignments : 2492264                                              ||
# ||    Successfully assigned alignments : 1877129 (75.3%)                      ||
# ||    Running time : 0.71 minutes                                             ||
# ||                                                                            ||
# ||                                                                            ||
# \\============================================================================//
```

