---
title: "DTU analysis of sequins data"
author: "Xueyi Dong"
date: "20/03/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

First, read in the data and correct counts by the tracking file

```{r}
data <- read.csv("/stornext/General/data/user_managed/grpu_mritchie_1/SCmixology/Mike_seqin/20200228_YPRDP_2xsequin_mixAB/FLTSA_output/transcript_count.csv", stringsAsFactors = FALSE)
anno <- read.delim("/stornext/General/data/user_managed/grpu_mritchie_1/SCmixology/Mike_seqin/annotations/rnasequin_isoforms_2.4.tsv", sep = "\t", stringsAsFactors = FALSE)
tracking <- read.table("comp.tracking", stringsAsFactors = FALSE)
info <- strsplit2(tracking$V5, "|", fixed=TRUE)
tracking$transcript <- sub("transcript:", "", info[,2])
tracking$transcript_anno <- strsplit2(tracking$V3, "|", fixed=TRUE)[,2]
tracking.eq <- tracking[tracking$V4=="=",]
m <- match(data$transcript_id, tracking.eq$transcript)
data$transcript_anno <- tracking.eq$transcript_anno[m]
data$transcript_anno[is.na(data$transcript_anno)] <- data$transcript_id[is.na(data$transcript_anno)]
```

Calculate correlation to know which is A and which is B

```{r}
m <- match(anno$NAME, data$transcript_anno)
cor(data[,3][m], anno[,3], us="complete.obs")
cor(data[,3][m], anno[,4], us="complete.obs")
cor(data[,4][m], anno[,3], us="complete.obs")
cor(data[,4][m], anno[,4], us="complete.obs")
cor(data[,5][m], anno[,3], us="complete.obs")
cor(data[,5][m], anno[,4], us="complete.obs")
cor(data[,6][m], anno[,3], us="complete.obs")
cor(data[,6][m], anno[,4], us="complete.obs")
```

sample 1: mix B
sample 2: mix B
sample 3: mix A
sample 4: mix A

```{r}
par(mfrow=c(2,2))
plot(data[,3][m], anno[,4], log="xy", 
     xlab = "count", ylab="theoretiacl abundance", main="sample 1, mix B")
plot(data[,4][m], anno[,4], log="xy", 
     xlab = "count", ylab="theoretiacl abundance", main="sample 2, mix B")
plot(data[,5][m], anno[,3], log="xy", 
     xlab = "count", ylab="theoretiacl abundance", main="sample 3, mix A")
plot(data[,6][m], anno[,3], log="xy", 
     xlab = "count", ylab="theoretiacl abundance", main="sample 4, mix A")
```


OK. Now prepare for DRIMSeq.

```{r}
library(DRIMSeq)
counts <- data[,c(2, 7, 3, 4, 5, 6)]
colnames(counts)[2] <- "feature_id"
samples <- data.frame(
  sample_id = colnames(counts)[3:6],
  group = c("B", "B", "A", "A")
)
# resolve duplicate
for(i in which(duplicated(counts$feature_id))){
  j = which(counts$feature_id == counts$feature_id[i])[1]
  counts[j,3:6] <- counts[j, 3:6] + counts[i, 3:6]
}
counts <- counts[!duplicated(counts$feature_id),]

d <- dmDSdata(counts=counts, samples = samples)
plotData(d)
d <- dmFilter(d, min_samps_gene_expr = 4, min_samps_feature_expr = 2, min_gene_expr = 10, min_feature_expr = 10, run_gene_twice = T)
plotData(d)
```

```{r}
design <- model.matrix(~group, data=DRIMSeq::samples(d))
colnames(design) <- sub("group", "", colnames(design))
set.seed(1904)
d <- suppressWarnings(dmPrecision(d, design, BPPARAM = BiocParallel::MulticoreParam(16)))

plotPrecision(d)
```

Check precision value.

Note: dispersion = 1 / (1 + precision)

```{r}
head(mean_expression(d))
common_precision(d)
head(genewise_precision(d))
```


## proportion estimation

```{r}
d <- suppressWarnings(dmFit(d, design = design, verbose=1, BPPARAM = BiocParallel::MulticoreParam(16)))
head(proportions(d))
```
## testing for differential transcript usage

Test results on gene level

```{r}
d <- dmTest(d, coef = "B")
head(results(d))
plotPValues(d)
```

Testing results on feature level (not that important for us)

```{r}
head(results(d, level = "feature"))
plotPValues(d, level = "feature")
```

Two-stage test (code copied from DRIMSeq vignette)

```{r}
res <- results(d)
res.txp <- results(d, level = "feature")
library(stageR)
## Assign gene-level pvalues to the screening stage
pScreen <- res$pvalue
# strp <- function(x) substr(x,1,18)
names(pScreen) <- results(d)$gene_id
## Assign transcript-level pvalues to the confirmation stage
pConfirmation <- matrix(res.txp$pvalue, ncol = 1)
rownames(pConfirmation) <- res.txp$feature_id
## Create the gene-transcript mapping
tx2gene <- res.txp[,c("feature_id", "gene_id")]
# for (i in 1:2) tx2gene[,i] <- strp(tx2gene[,i])
## Create the stageRTx object and perform the stage-wise analysis
stageRObj <- stageRTx(pScreen = pScreen, pConfirmation = pConfirmation,
pScreenAdjusted = FALSE, tx2gene = tx2gene)
stageRObj <- stageWiseAdjustment(object = stageRObj, method = "dtu",
alpha = 1)

suppressWarnings({
  drim.padj <- getAdjustedPValues(stageRObj, order=TRUE,
                                  onlySignificantGenes=TRUE)
})
head(drim.padj)
# write.table(drim.padj, file = "DTUpadj.txt", sep = "\t", row.names = FALSE)
```

drimseq proportion vs real proportion

calculate annotation proportion.
```{r}
geneanno <- read.delim("/stornext/General/data/user_managed/grpu_mritchie_1/SCmixology/Mike_seqin/annotations/rnasequin_genes_2.4.tsv", sep = "\t", stringsAsFactors = FALSE)
proportion <- data.frame(txID = anno$NAME)
tx2gene <- limma::strsplit2(proportion$txID, "_")
tx2gene <-  paste(tx2gene[,1], tx2gene[,2], sep="_")
proportion$geneID <- tx2gene
m <- match(proportion$geneID, geneanno$NAME)
proportion$MIX_A <- anno$MIX_A / geneanno$MIX_A[m]
proportion$MIX_B <- anno$MIX_B / geneanno$MIX_B[m]

m <- match(proportion$txID, proportions(d)[,2])
cor(proportion$MIX_B, proportions(d)[,3][m], use="complete.obs")
cor(proportion$MIX_A, proportions(d)[,5][m], use="complete.obs")
```

```{r}
par(mfrow=c(1,2))
plot(proportion$MIX_B, proportions(d)[,3][m], main="mix B",
     xlab = "expected proportion", ylab = "estimated proportion")
plot(proportion$MIX_A, proportions(d)[,5][m], main="mix A",
     xlab = "expected proportion", ylab = "estimated proportion")
```

run diffSplice

```{r}
library(edgeR)
library(limma)
y <- DGEList(counts = as.matrix(DRIMSeq::counts(d)[,c(-1, -2)]), samples = DRIMSeq::samples(d))
rownames(y) <- counts(d)$feature_id
y <- calcNormFactors(y)
y$samples
cpm <- cpm(y, log=T)
plotMDS(cpm, col=as.numeric(as.factor(group)))
```

```{r fit}
v <- voom(y, design, plot=T)
fit <- lmFit(v,design)
efit <- eBayes(fit)
dt <- decideTests(efit, p.value = 0.25)
summary(dt)
```

```{r}
geneid <- counts(d)$gene_id
featureid <- counts(d)$feature_id

ex <- diffSplice(efit, geneid = geneid, exonid=featureid)
ts <- topSplice(ex,  number=Inf)
ts.tx <- topSplice(ex, test = "t", number = Inf)

head(ts)
table(ts$FDR < 0.05)
```

Now deal with anno, find out control genes

```{r}
table(anno$MIX_A==anno$MIX_B)
proportion$FC <- proportion$MIX_A / proportion$MIX_B
proportion$isChanges <- FALSE
proportion$isChanges[abs(proportion$FC-1) > 0.001] <- TRUE
```

check DRIMSeq results

```{r}
res.tx <- results(d, level = "feature")
res.tx$isChanges <- proportion$isChanges[match(res.tx$feature_id, proportion$txID)]
ggplot(na.omit(res.tx)) +
  geom_boxplot(aes(x=isChanges, y=adj_pvalue)) +
  theme_bw() +
  ggtitle("DRIMSeq transcript-level FDR") +
  labs(x = "isProportionChange")
ggplot(na.omit(res.tx)) +
  geom_boxplot(aes(x=isChanges, y=pvalue)) +
  theme_bw() +
  ggtitle("DRIMSeq transcript-level p value") +
  labs(x = "isProportionChange")
```

```{r}
res.tx$logFC <-log( proportion$FC[match(res.tx$feature_id, proportion$txID)])
ggplot(res.tx, aes(x=logFC, y=adj_pvalue)) +
  geom_point() +
  theme_bw() +
  ggtitle("DRIMSeq")
```


And check diffSplice results

```{r}
ts.tx$isChanges <- proportion$isChanges[match(ts.tx$ExonID, proportion$txID)]
ggplot(na.omit(ts.tx), aes(x=isChanges, y=FDR)) +
  geom_boxplot() +
  theme_bw() +
  ggtitle("diffSplice transcript-level FDR") +
  labs(x = "isProportionChange")
  
```

```{r}
ts.tx$logFC <- log(proportion$FC[match(ts.tx$ExonID, proportion$txID)])

ggplot(ts.tx, aes(x=logFC, y=FDR)) +
  geom_point() +
  theme_bw() +
  ggtitle("diffSplice")
```



```{r}
save.image("DTUsequins.RData")
```

