---
title: "Untitled"
author: "Xueyi Dong"
date: "03/05/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
dir <- "/home/users/allstaff/mritchie/mattlab/long_RNA_benchmark/Illumina_April19"
libs <- c(paste(1:2, "-H1975-", 1:2, "_S", 1:2, sep=""), paste(6:7, "-HCC-", 1:2, "_S", 6:7, sep=""))
bam <- paste(dir, "/results/subread_sequins/", libs, ".sorted.bam", sep="")
gtf <- "~/annotation/sequins/rnasequin_annotation_2.4.gtf"
```

```{r}
library(Rsubread)
```

```{r}
fc <- featureCounts(bam, 
                    annot.ext = gtf, 
                    isGTFAnnotationFile = TRUE, 
                    isPairedEnd = TRUE, 
                    nthreads=8)
saveRDS(fc, file="counts_subread_Illumina_April19.RDS")
# fc <- readRDS("/stornext/Home/data/allstaff/d/dong.x/analysis/2020/long_read_benchmark/counts_Illumina_April19.RDS")
```

```{r}
library(edgeR)
library(limma)

x <- DGEList(counts=fc$counts, genes=fc$annotation, group=factor(c(rep("H1975", 2), rep("HCC", 2))))
dim(x)
colnames(x) <- libs
x$samples
```


```{r}
filter <- filterByExpr(x)
table(filter)
x <- x[filter,]
x <- calcNormFactors(x)
x$samples
cpm <- cpm(x, log=TRUE)
pdf("MDS.pdf", height = 5)
plotMDS(cpm, col=c("red", "blue")[as.numeric(x$samples$group)], pch = 1)
legend("topright", c("H1975", "HCC827"), text.col=c("red", "blue"), pch = 1, col = c("red", "blue"))
dev.off()
# par(mfrow=c(1,2))
# plotMDS(cpm[,1:5], labels=libs[1:5])
# plotMDS(cpm[,6:10], labels=libs[6:10])
```

```{r}
summary(x$counts)
```

```{r}
hist(x$counts)
```

Plot CPM estimate vs expect

```{r}
anno <- read.table("/wehisan/home/allstaff/d/dong.x/annotation/sequins/rnasequin_genes_2.4.tsv", header = TRUE, stringsAsFactors = FALSE)
anno$logFC <- log(anno$MIX_B / anno$MIX_A)
cpm <- cpm(x)
m <- match(anno$NAME, rownames(cpm))
annoData <- cbind(anno, cpm[m,])
colnames(annoData)[6:9] <- paste0("mix", rep(c("A", "B"), c(2, 2)), c(1:2, 1:2))
annoData <- data.table::melt(annoData, id.vars = 1:5)
annoData$exp <-annoData$MIX_A
annoData$exp[annoData$variable %in% c("mixB1", "mixB2", "mixB3", "mixB4", "mixB5")] <- annoData$MIX_B[annoData$variable %in% c("mixB1", "mixB2", "mixB3", "mixB4", "mixB5")]
colnames(annoData)[6] <- "sample"
annoData$sample <- factor(annoData$sample, levels = paste0("mix", rep(c("A", "B"), c(2, 2)), c(1:2, 1:2)))
cor(annoData$exp, annoData$value, use = "complete.obs")

library(ggplot2)
library(scales)
pdf("plots/GeneCPMvsAnno.pdf", height = 5, width = 8)
ggplot(annoData, aes(x=exp, y=value, colour=sample)) +
  geom_point(position = position_jitterdodge(dodge.width = 0.1)) +
  theme_bw() +
  scale_x_continuous(trans = "log10", labels = comma) +
  scale_y_continuous(trans = "log10", labels=comma, limits=c(0.3, 250000)) +
  labs(x = "Expected abundance", y = "CPM") +
  scale_colour_manual(values = c("#869700", "#D12300", "#071535", "#54B5E1"))+
  theme(text = element_text(size = 20)) +
    annotate(geom="text", x=300, y=230000, label=expression(paste(rho, "=0.80")), size=8) 
dev.off()
```

```{r}
design <- model.matrix((~x$samples$group))
design

```



```{r}
v <- voom(x, design=design, plot=TRUE, save.plot=TRUE)
fit <- lmFit(v, design = design)
# fit <- contrasts.fit(fit, contrasts = cont)
efit <- eBayes(fit)
dt <- decideTests(efit)
summary(dt)
```

```{r}
tt <- topTable(efit, coef=2, number=Inf)
anno <- read.table("/wehisan/home/allstaff/d/dong.x/annotation/sequins/rnasequin_genes_2.4.tsv", header = TRUE, stringsAsFactors = FALSE)
anno$logFC <- log(anno$MIX_B / anno$MIX_A)
tt$logFC_expected <- anno$logFC[match(rownames(tt), anno$NAME)]
tt$LENGTH <- anno$LENGTH[match(rownames(tt), anno$NAME)]
cor(tt$logFC, tt$logFC_expected)
lfc.lm <-lm(tt$logFC ~ tt$logFC_expected)
summary(lfc.lm)
```

```{r}
write.table(tt, file = "topTable.txt", sep = "\t")
```

