---
title: "DEU analysis for sequins Illumina data"
author: "Xueyi Dong"
date: "10/11/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(Rsubread)
library(edgeR)
library(limma)
```

```{r}
dir <- "/home/users/allstaff/mritchie/mattlab/long_RNA_benchmark/Illumina_April19"
libs <- c(paste(1:5, "-H1975-", 1:5, "_S", 1:5, sep=""), paste(6:10, "-HCC-", 1:5, "_S", 6:10, sep=""))
bam <- paste(dir, "/results/hisat2_sequins/", libs, ".hisat2.bam", sep="")
gtf <- "~/annotation/sequins/rnasequin_annotation_2.4.gtf"
```

```{r}
fc_exon <- featureCounts(bam, 
                    annot.ext = gtf, 
                    isGTFAnnotationFile = TRUE, 
                    GTF.featureType = "exon",
                    GTF.attrType = "gene_id",
                    isPairedEnd = TRUE, 
                    allowMultiOverlap = TRUE,
                    useMetaFeatures = FALSE,
                    nthreads=8)
```
```{r}
saveRDS(fc_exon, "conuts_exon.RDS")
```

organize object

```{r}
ind = match(fc_exon$anno[,1], fc_exon$annotation[,1])
x = DGEList(counts=fc_exon$counts)
x$genes = fc_exon$annotation[ind,]
x$samples$group <- rep(c("mixA", "mixB"), c(5, 5))
```

filter lowly expressed exons and normalize

```{r}
filter <- filterByExpr(x)
table(filter)
x <- x[filter,]
x <- calcNormFactors(x)
```

Fit model

```{r}
design <- model.matrix(~x$samples$group)
v <- voom(x, design, plot=TRUE)
fit <- lmFit(v, design=design)
efit <- eBayes(fit)
```

run DEU analysis

```{r}
ds <- diffSplice(efit, geneid="GeneID")
ts <- topSplice(ds, coef=2, number=Inf)
```

explore results

```{r}
sum(ts$FDR<0.05)
head(ts)
```

