---
title: "DTU analysis for sequins Illumina data salmon counts"
author: "Xueyi Dong"
date: "23/11/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Organize transcript count data and annotation

```{r}
datadir="/stornext/General/data/user_managed/grpu_mritchie_1/long_RNA_benchmark/Illumina_April19/results/salmon"
samples <- c("1-H1975-1_S1","2-H1975-2_S2", "6-HCC-1_S6", "7-HCC-2_S7")
quant <- file.path(datadir, samples, "quant.sf")
library(tximport)
txi <- tximport(quant, type="salmon", txOut=TRUE, countsFromAbundance = "scaledTPM")
library(GenomicFeatures)
gtf <- "/stornext/General/data/user_managed/grpu_mritchie_1/SCmixology/Mike_seqin/20200228_YPRDP_2xsequin_mixAB/flames/isoform_annotated.filtered.gff3"
txdb <- makeTxDbFromGFF(gtf)
txdf <- select(txdb, keys(txdb, "GENEID"), "TXNAME", "GENEID") 
tab <- table(txdf$GENEID) 
txdf$ntx <- tab[match(txdf$GENEID, names(tab))]
```


## DTU analysis using DRIMSeq

```{r}
library(DRIMSeq)
samples <- data.frame(
  sample_id = make.names(samples),
  group = c("A", "A", "B", "B"),
  stringsAsFactors = FALSE
)
counts <- as.data.frame(txi$counts, stringAsFactors = FALSE)
counts <- counts[match(txdf$TXNAME, rownames(counts)),]
counts <- cbind(txdf$GENEID, txdf$TXNAME, counts)
colnames(counts) <- c("gene_id", "feature_id", samples$sample_id)
d <- dmDSdata(counts = counts, samples = samples)
d
```


### count vs annotation plot

For long read data, we used CPM. But for short read data, use TPM to adjust length bias.

```{r}
anno <- read.delim("/stornext/General/data/user_managed/grpu_mritchie_1/SCmixology/Mike_seqin/annotations/rnasequin_isoforms_2.4.tsv", sep = "\t", stringsAsFactors = FALSE)
m <- match(anno$NAME, as.character(counts$feature_id))
annoData <- cbind(anno, counts[m, 3:6])
# colnames(annoData)[5:8] <- paste0("mix", c("A1", "A2", "B1", "B2"))
annoData <- data.table::melt(annoData, id.vars = 1:4)
annoData$exp <-annoData$MIX_A
annoData$exp[annoData$variable %in% c("6-HCC-1_S6", "7-HCC-2_S7")] <- 
  annoData$MIX_B[annoData$variable %in% c("6-HCC-1_S6", "7-HCC-2_S7")]
colnames(annoData)[5] <- "sample"
# annoData$sample <- factor(annoData$sample, levels = c("mixA1", "mixA2", "mixB1", "mixB2"))
cor(annoData$exp, annoData$value, use = "complete.obs")

library(ggplot2)
library(scales)
pdf("plots/countsVsAnno.pdf", height = 5, width = 8)
ggplot(annoData, aes(x=exp, y=value, colour=sample)) +
  geom_point() +
  theme_bw() +
  scale_x_continuous(trans = "log10", labels = comma) +
  scale_y_continuous(trans = "log10", labels=comma) +
  labs(x = "Expected abundance", y = "TPM") +
  scale_colour_manual(values = c("#869700", "#D12300", "#071535", "#54B5E1"))+
  theme(text = element_text(size = 20)) +
  annotate(geom="text", x=100, y=3000000, label=expression(paste(rho, "=0.86")), size=8) 
dev.off()
```




```{r}
d <- dmFilter(d, min_samps_gene_expr = 2, min_samps_feature_expr = 2, min_gene_expr = 10, min_feature_expr = 10, run_gene_twice = T)
plotData(d)
```

```{r}
design <- model.matrix(~group, data=DRIMSeq::samples(d))
colnames(design) <- sub("group", "", colnames(design))
set.seed(1904)
d <- suppressWarnings(dmPrecision(d, design, BPPARAM = BiocParallel::MulticoreParam(8)))

plotPrecision(d)
```

```{r}
d <- suppressWarnings(dmFit(d, design = design, verbose=1, BPPARAM = BiocParallel::MulticoreParam(16)))
head(proportions(d))
```

```{r}
d <- dmTest(d, coef = "B")
head(results(d))
plotPValues(d)
```

```{r}
head(results(d, level = "feature"))
plotPValues(d, level = "feature")
```

stageR analysis

```{r}
res <- results(d)
res.txp <- results(d, level = "feature")
no.na <- function(x) ifelse(is.na(x), 1, x) 
res$pvalue <- no.na(res$pvalue) 
res.txp$pvalue <- no.na(res.txp$pvalue)

library(stageR)
## Assign gene-level pvalues to the screening stage
pScreen <- res$pvalue
# strp <- function(x) substr(x,1,18)
names(pScreen) <- results(d)$gene_id
## Assign transcript-level pvalues to the confirmation stage
pConfirmation <- matrix(res.txp$pvalue, ncol = 1)
rownames(pConfirmation) <- res.txp$feature_id
## Create the gene-transcript mapping
tx2gene <- res.txp[,c("feature_id", "gene_id")]
# for (i in 1:2) tx2gene[,i] <- strp(tx2gene[,i])
## Create the stageRTx object and perform the stage-wise analysis
stageRObj <- stageRTx(pScreen = pScreen, pConfirmation = pConfirmation,
pScreenAdjusted = FALSE, tx2gene = tx2gene)
stageRObj <- stageWiseAdjustment(object = stageRObj, method = "dtu",
alpha = 0.05)

suppressWarnings({
  drim.padj <- getAdjustedPValues(stageRObj, order=TRUE,
                                  onlySignificantGenes=FALSE)
})
head(drim.padj)
# write.table(drim.padj, file = "DTUpadj.txt", sep = "\t", row.names = FALSE)
```

## run diffSplice

```{r}
library(edgeR)
library(limma)
y <- DGEList(counts = as.matrix(DRIMSeq::counts(d)[,c(-1, -2)]), samples = DRIMSeq::samples(d))
rownames(y) <- counts(d)$feature_id
y <- calcNormFactors(y)
y$samples
cpm <- cpm(y, log=T)
plotMDS(cpm, col=as.numeric(as.factor(y$samples$group)))
```

```{r fit}
v <- voom(y, design, plot=T)
fit <- lmFit(v,design)
efit <- eBayes(fit)
dt <- decideTests(efit, p.value = 0.25)
summary(dt)
```

```{r}
geneid <- counts(d)$gene_id
featureid <- counts(d)$feature_id

ex <- diffSplice(efit, geneid = geneid, exonid=featureid)
ts <- topSplice(ex,  number=Inf)
ts.tx <- topSplice(ex, test = "t", number = Inf)
ts.f <- topSplice(ex, test = "F", number = Inf)

head(ts)
table(ts$FDR < 0.05)
```



Two-stage test for diffSplice

```{r}

library(stageR)

pScreen <- ts$P.Value
pScreen2 <- ts.f$P.Value
names(pScreen) <- ts$GeneID
names(pScreen2) <- ts.f$GeneID
pConfirmation <- matrix(ts.tx$P.Value, ncol = 1)
rownames(pConfirmation) <- ts.tx$ExonID

tx2gene <- ts.tx[,1:2]

stageRObj2 <- stageRTx(pScreen = pScreen, pConfirmation = pConfirmation,
pScreenAdjusted = FALSE, tx2gene = tx2gene)
stageRObj2 <- stageWiseAdjustment(object = stageRObj2, method = "dtu",
alpha = 0.05)
stageRObj3 <- stageRTx(pScreen = pScreen2, pConfirmation = pConfirmation,
pScreenAdjusted = FALSE, tx2gene = tx2gene)
stageRObj3 <- stageWiseAdjustment(object = stageRObj3, method = "dtu",
alpha = 0.05)
suppressWarnings({
  ds.padj <- getAdjustedPValues(stageRObj2, order=TRUE,
                                  onlySignificantGenes=FALSE)
  ds.padj.f <- getAdjustedPValues(stageRObj3, order=TRUE,
                                  onlySignificantGenes=FALSE)
})
head(ds.padj)
head(ds.padj.f)

```

## check results
question: what should I do? Compare with long read? 
Do I need to make a plot separately for short read?
Maybe a scatter plot compare t btw long and short?

save some results to table
```{r}
# gene level diffSplice result (simes)
write.table(ts, "topSplice_simes.tsv", sep = "\t")
# drimseq+stager results
write.table(drim.padj, "drimseq_stager.tsv", sep = "\t")
# drimseq gene
write.table(res, "drimseq_gene.tsv", sep = "\t")
# drimseq proportion
write.table(proportions(d), "drimseq_proportion.tsv", sep="\t")

```

```{r}
plotProportions(d, gene_id = "R1_32", group_variable = "group", plot_type = "barplot") + 
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1), 
        text = element_text(size=12), 
        legend.text=element_text(size=10))
```

