---
title: "QC plots"
author: "Xueyi Dong"
date: "30/09/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

Plots to be made:

* Read length
* Alignment reads/sample
* read quality metrics
* QC metrics (probably by some other software)

# Make plots

Tried **nanoQC**. Results not as good as expected.

Also Tried **AlignQC**. Very time consuming. Didn't see any output and I don't know why.

Now try **nanoR**.

```{r}
library(NanoR)
pathSummary <- "/wehisan/general/academic/seq_data/quentin/Nanopore/Smchd1-NSC-cDNA/Xueyi/basecalled/sequencing_summary.txt.gz"
pathFq <- "/wehisan/general/academic/seq_data/quentin/Nanopore/Smchd1-NSC-cDNA/Xueyi/demultiplex/filtered/"
List <- NanoPrepareG(DataSummary=pathSummary, DataFastq=pathFq)
Table <- NanoTableG(NanoGList = List, DataOut="NanoRResult/rep1_pass", GCC=FALSE)
```

```{r}
NanoStatsG(NanoGList = List, NanoGTable = Table, DataOut="NanoRResult/rep1")
```

```{r}
pathRep2 <- "/wehisan/general/academic/seq_data/quentin/Nanopore/Smchd1-NSC-cDNA_rep2/data/basecalled"
List2 <- NanoPrepareG(DataSummary=pathRep2, DataFastq=pathRep2, Cores = 1, Label="Exp")
Table2 <- NanoTableG(NanoGList = List2, DataOut="NanoRResult/rep2", GCC=FALSE)
NanoStatsG(NanoGList = List2, NanoGTable = Table2, DataOut="NanoRResult/rep2")
```

